File: framework/languages/java/e2e-kotlin/SpotService/Client/src/main/kotlin/systems/zlink/e2e/kotlin/spotservice/client/scenarios/SmB1Scenario.kt
package systems.zlink.e2e.kotlin.spotservice.client.scenarios

import systems.zlink.e2e.kotlin.spotservice.client.support.ActorSessionScenarioContext
import systems.zlink.e2e.kotlin.spotservice.client.support.ensure

internal object SmB1Scenario {
    suspend fun run(context: ActorSessionScenarioContext) {
        context.connectAndAuthenticate()
        ensure(context.auth.actorId == context.actorId, "SM-B1 auth actor mismatch")
        ensure(context.auth.boundCount == 1, "SM-B1 bound actor count mismatch")

        context.requestEntryEcho()
        ensure(
            context.entryReply.value == "entry:entry-echo",
            "SM-B1 entry actor request mismatch: ${context.entryReply.value}"
        )

        context.joinRoom()
        ensure(context.joined.spotRid == "actor-room-a", "SM-B1 joined spot mismatch")

        context.requestFirstUserEchoAndCheckUnboundIsolation()
        ensure(context.userReply1.spotRid == "actor-room-a", "SM-B1 user actor spot mismatch")
        ensure(
            context.userReply1.value == "user:user-echo-1",
            "SM-B1 user actor request mismatch: ${context.userReply1.value}"
        )

        println("scenario SM-B1 passed")
    }
}

File: framework/languages/node/e2e/SpotService/Client/Scenarios/sm-b1-scenario.ts
// SM-B1: 같은 node의 User Spot으로 Join한다 시나리오를 검증한다.
import {
  zlinkStreamConnectorFactory,
  zlinkStreamJsonCodec,
  ZlinkStreamDispatchMode
} from '@zlink-systems/stream-connector';
import type {
  ActorPingRes,
  ActorPingReq,
  AuthRes,
  AuthReq,
  EvidenceWaitReq
} from '../../Shared/messages';
import type { ClientOptions } from '../Support/client-options';
import { postJson } from '../../../http-client';
import { ensure } from '../Support/scenario-assert';

export async function runSmB1(options: ClientOptions): Promise<void> {
  const actorId = `actor-sm-b1-local-${Date.now()}`;
  const client = zlinkStreamConnectorFactory.create({
    endpoint: options.sessionAStreamEndpoint,
    codec: zlinkStreamJsonCodec,
    dispatchMode: ZlinkStreamDispatchMode.Immediate,
    heartbeat: { enabled: false },
    waitTimeoutMs: 10000
  });
  await client.connect();
  try {
    const auth = await client
      .request({
        actorId,
        displayName: 'local actor',
        nodeRid: 'play-a'
      } satisfies AuthReq)
      .packetName('AuthReq')
      .timeout(5000)
      .submit<AuthRes>();
    ensure(auth.actorId === actorId && auth.nodeRid === 'play-a', 'SM-B1 auth reply mismatch.');

    const pingMsg = await client
      .request({ value: 'b1' } satisfies ActorPingReq)
      .packetName('ActorPingReq')
      .timeout(5000)
      .submit<ActorPingRes>();
    ensure(pingMsg.actorId === actorId, 'SM-B1 actor reply mismatch.');
    ensure(pingMsg.nodeRid === 'play-a', 'SM-B1 local node mismatch.');

    const expectedEvidence = [
      `entry-created|rid=play-a|actor=${actorId}`,
      `entry-joined|rid=play-a|actor=${actorId}`
    ];
    const evidence = await postJson<string[]>(options.playAUrl, '/evidence/wait', {
      containsAll: expectedEvidence,
      timeoutMilliseconds: 10000
    } satisfies EvidenceWaitReq);
    ensure(
      expectedEvidence.every((expected) => evidence.some((line) => line.includes(expected))),
      'SM-B1 evidence mismatch.'
    );
  } finally {
    await client.close();
  }

  console.log('scenario SM-B1 passed');
}
